src/Microsoft.AspNetCore.SystemWebAdapters/HttpFileCollection.cs

```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Http;
+
+namespace System.Web;
+
+[SuppressMessage("Design", "CA1010:Collections should implement generic interface", Justification = Constants.ApiFromAspNet)]
+public sealed class HttpFileCollection : NameObjectCollectionBase
+{
+    private string[]? _keys;
+
+    public HttpFileCollection(IFormFileCollection files)
+    {
+        FormFiles = files;
+
+        IsReadOnly = true;
+    }
+
+    internal IFormFileCollection FormFiles { get; }
+
+    [SuppressMessage("Performance", "CA1819:Properties should not return arrays", Justification = Constants.ApiFromAspNet)]
+    public string[] AllKeys => _keys ??= GetKeys();
+
+    public override int Count => FormFiles.Count;
+
+    [Obsolete("Retrieving Keys is not supported on .NET 6+. Please use the enumerator instead.")]
+#pragma warning disable CS0809 // Obsolete member overrides non-obsolete member
+    public override KeysCollection Keys => throw new PlatformNotSupportedException();
+#pragma warning restore CS0809 // Obsolete member overrides non-obsolete member
+
+    public override IEnumerator GetEnumerator() => AllKeys.GetEnumerator();
```

Comment:
Should this really be returning keys rather than files?

Reply:
Weirdly, that's what the NameObjectCollectionBase does. You iterate through to get the keys, and then you end up having to index into it by string